xla/client/lib/math.cc

```diff
@@ -1175,10 +1175,130 @@ XlaOp Acos(XlaOp x) {
   });
 }
 
+// hypot(x, y) = max(|x|, |y|) * sqrt(1 + (min(|x|, |y|) / max(|x|, |y|))^2)
+static XlaOp Hypot(XlaOp x, XlaOp y) {
+  auto ax = Abs(x);
+  auto ay = Abs(y);
+  auto mx = Select(Gt(ax, ay), ax, ay);
+  auto mn = Select(Gt(ax, ay), ay, ax);
+  auto mnomx = Div(mn, mx);
+  return Mul(mx, Sqrt(Add(ScalarLike(x, 1), Mul(mnomx, mnomx))));
+}
+
 // asin(x) = 2 * atan(x / (1 + sqrt(1 - x^2)))
 XlaOp Asin(XlaOp x) {
-  return ScalarLike(x, 2.0) *
-         Atan2(x, ScalarLike(x, 1.0) + Sqrt(ScalarLike(x, 1.0) - x * x));
+  XlaBuilder* b = x.builder();
+  auto do_it = [&](XlaOp z) -> absl::StatusOr<XlaOp> {
+    TF_ASSIGN_OR_RETURN(auto shape, b->GetShape(z));
+    if (primitive_util::IsComplexType(shape.element_type())) {
+      // According to "Implementing the complex arcsine and arccosine
+      // functions using exception handling" by Hull et al
+      // [https://dl.acm.org/doi/10.1145/275323.275324], complex
+      // arcsin function can be defined as
+      //   arcsin(z) = arcsin(x/a) + sign(y; x) * I * log(a + sqrt(a*a-1))
```

Comment:
I'm assuming this is for `z = x + I*y`, yes?

Reply:
yes. I'll extend the comment accordingly.

Reply:
Fixed, an updated asin algorithm description is in math_impl.h